subsys/bluetooth/host/conn.c

```diff
@@ -789,10 +852,13 @@ void bt_conn_process_tx(struct bt_conn *conn)
 		return;
 	}
 
-	/* Get next ACL packet for connection */
-	buf = net_buf_get(&conn->tx_queue, K_NO_WAIT);
+	/* Get next ACL packet for connection. The buffer will only get dequeued
+	 * if there is a free controller buffer to put it in.
+	 */
+	buf = k_fifo_peek_head(&conn->tx_queue);
```

Comment:
The problem with this is that you're skipping the fragment handling that `net_buf_get()` takes care of. We don't (AFAIK) currently take advantage of the net_buf fragment feature in the Bluetooth subsystem, but at least we've used the proper net_buf APIs until now that should keep the code working if we did use fragments.

At the very least, in addition to asserting that buf != NULL, you should also `BT_ASSERT((buf->flags & NET_BUF_FRAGS) == 0);`. Though even that's pretty unclean, since these are really net_buf-internal features.

A cleaner way would be to introduce a new `net_buf_peek()` API that does the same as `net_buf_get()` but also fetches any pending fragments.

Yet another, perhaps more long term solution, would be to finally move the net_buf fragments list into it's own independent list that's not mangled together with the linked list pointer as part of the same union (which is the reason for the NET_BUF_FRAGS flag and why fragments are placed as independent objects into the FIFO). Currently the net_buf definition starts off like this:
```c
struct net_buf {
	union {
		/** Allow placing the buffer into sys_slist_t */
		sys_snode_t node;

		/** Fragments associated with this buffer. */
		struct net_buf *frags;
	};
```

The penalty of moving `frags` out from the union would be that the struct size increases by the size of a pointer, but the benefit would be that we don't have this hassle of serializing & deserializing fragments whenever buffers pass through LIFO or FIFO objects.

Reply:
> At the very least, in addition to asserting that buf != NULL, you should also `BT_ASSERT((buf->flags & NET_BUF_FRAGS) == 0);`. Though even that's pretty unclean, since these are really net_buf-internal features.

Oh, and in addition to the assert, you need to explicitly set `buf->frags = NULL` since the k_fifo implementation leaves the linked list pointer untouched when removing from the list.

Also see https://github.com/zephyrproject-rtos/zephyr/pull/38830#pullrequestreview-763305021 and related discussion in the PR.

Reply:
is that really a problem since I'm using the `_get` before passing it to `bt_send`? That takes care of the fragments, and the peek is a non-destructive operation afaik.

Reply:
At the very least you need to clear the `buf->frags` pointer since `k_fifo_peek_head()` doesn't do that for you. You have the `EIO` branch below where e.g. `net_buf_unref()` is called, and `net_buf_unref()` will also unref the frags pointer if it is non-NULL. Same goes for any other net_buf API calls you might do on the buffer before calling `net_buf_get()`

Reply:
Actually, I just realised that the `EIO` branch is probably buggy in this regard, i.e. of `send_buf()` returns error before getting to the `net_buf_get()` phase, you have a pointer to something that's still in the FIFO and you don't own a reference to, yet you still call `net_buf_unref()` on it.

Reply:
This `frags` business is really backwards IMO, why do we even use it? seems like a linked-list implementation that we then put on another linked-list (the queue)?
Then we need all those safe APIs to deal with that instead of being able to just use the fifo/queue apis directly.
anyway for the EIO case, the buffer has already been consumed by the `_get`.

Reply:
> Why do we want to clear the frags ptr again? isn't anything in frags actually part of the current buffer?

Yes, but only if those buffers were removed from the FIFO that they were passed through. If you use k_fifo APIs directly you may end up with a frags pointing at buffers that are still inside the FIFO, i.e. it's the FIFO that owns their references, not the parent buffer. Since we don't use frags (it's a feature added by the networking stack folks) we could for now just assert that frags were not used and then explicitly set the buf->frags pointer to NULL.

Reply:
Anyway, as I said, really the cleanest way out of this would be to detach the frags from being in a union together with the linked list pointer in net_buf. But that's also considerably more work, more intrusive (affecting things outside of Bluetooth) and also comes with an increased memory footprint.

Reply:
> This frags business is really backwards IMO, why do we even use it? seems like a linked-list implementation that we then put on another linked-list (the queue)?

You'll have to as the networking subsystem folks about that. I don't think they use FIFOs as much as we do, but then they do have an abstraction called net_pkt which has a pointer to the "top" buffer, which may then have a fragment list.

> anyway for the EIO case, the buffer has already been consumed by the _get.

Even then, this code looks pretty fragile, and could break if it's modified by someone not intimately familiar with net_buf internals. Anyway, if you make sure you always own a reference to the buffer when having a pointer to it (i.e. call net_buf_ref() after peek()) then the code becomes more future-proof.

Reply:
I also thought we weren't using frags (ie I had an assert doing just that before), but there seem to be some use of it in `l2cap.c` regarding SDU handling.
I'm not sure I'm reading the code correctly though, so I'd appreciate if you took a look just to double-check.

Reply:
One compromise I just thought of: maybe we could introduce a new CONFIG_NET_BUF_FRAGS option to enable/disable net_buf fragments support. Since we don't use/need it, we could keep it disabled, as long as you don't enable some other subsystem that depends on it (like the networking stack itself).

Reply:
> introduce a new CONFIG_NET_BUF_FRAGS option 

Well, that'd take care of the memory overhead that'd come with `frags` being outside of the union, but we'd still need to use the proper APIs so that our code doesn't break if the option is enabled.

Reply:
> Even then, this code looks pretty fragile, and could break if it's modified by someone not intimately familiar with net_buf internals.
I think this can be argued for the whole BT host. I know I've broken a bunch of stuff because of hidden assumptions like these.
I thought about moving the error handling somewhere else, but all the other places seemed messy as well. I'll at least add the referencing (and frags assert) as you said.

Reply:
> I also thought we weren't using frags (ie I had an assert doing just that before), but there seem to be some use of it in l2cap.c regarding SDU handling.

Right, I had forgotten about that. IIRC that's code for 6LoWPAN/IPSP when we receive buffers from the Networking subsystem that we should then send over L2CAP Connection-oriented Channels. In such a case the Net subsystem might feed us with fragment lists.

Reply:
It looks like the `l2cap.c` frags code stays within the L2CAP layer, i.e. when passing to ACL (conn->tx_queue) there are no fragment buffers anymore.

Reply:
thanks for looking into it. that was my initial understanding as well, just wanted to be sure, since we pass the buffers around and I'm not always sure what we do with those frags.

I will add the assert when we put the buffers in the queue, I think it's safer.

Reply:
Can I have it at both places? because good luck getting a stack trace of who put that bad buffer in if it hits when popping it rather than when pushing it in.

Reply:
Yes, that's fine. We could make the writing side a runtime check too, if we consider this the public API for that layer.

Reply:
> Oh, and in addition to the assert, you need to explicitly set buf->frags = NULL since the k_fifo implementation leaves the linked list pointer untouched when removing from the list.

I just tried and it didn't work. I think it's because `buf->frags` is in a union with the node ptr. 
So if I overwrite it with `NULL` while it is still on the queue, and it ends up that I don't pull it from the fifo (e.g. the sem_take fails), the fifo list is effectively broken.

Reply:
Right, good point. And this is another reason why using peak() together with the union is bad - you end up with the buffer in an inconsistent state until you actually remove it from from the FIFO.

Reply:
The short term solution would then probably be to document this clearly and try to make the code flow path from peek() to get() as short and obvious as possible.

Reply:
yep agreed. I will place a comment above the `peek` saying this.

Reply:
so I reverted the assert since #52760 got merged.